server/automated_tests_server.rb

```diff
@@ -43,15 +43,21 @@ def self.perform(markus_address, user_api_key, server_api_key, test_username, te
       start_time = Time.now
       Open3.popen3(run_command, pgroup: true) do |stdin, stdout, stderr, thread|
         pid = thread.pid
-        # mimic capture3 to read safely
-        stdout_thread = Thread.new { stdout.read }
-        stderr_thread = Thread.new { stderr.read }
+        stdin.close
+        # mimic capture3 to read safely and capture each line as it comes so that
+        # these threads don't hang or raise an error if we have to kill them early
+        stdout_thread = Thread.new { stdout.each { |line| output << line + $/} }
+        stderr_thread = Thread.new { stderr.each { |line| errors << line + $/} }
         if !thread.join(script['timeout']) # still running, let's kill the process group
           if test_username.nil?
             Process.kill('KILL', -pid)
           else
             Open3.capture3("sudo -u #{test_username} -- bash -c \"kill -KILL -#{pid}\"")
           end
+          # prepend errors with output up to when the timeout occured
+          unless output.empty?
+            errors = ["STDOUT BEFORE TIMEOUT OCCURED:", output, "ERRORS:", errors].join($/)
+          end
```

Comment:
1) Like above, string interpolation and \n are more readable.
2) Use TEST RESULTS instead of STDOUT, just for consistency with our messages on the Markus side.
3) It would be nice not to have ERRORS if errors.empty?